### ISSUE TITLE
Allow EF to call parameterized entity constructors

### ISSUE THREAD

**Issue by codepb** [aspnet/EntityFrameworkCore#10703]
When trying to create classes with only readonly properties, you currently have to use properties with backing fields to be able to construct with Entity Framework, and it is required to have an empty constructor. It would be preferable if Entity Framework could construct the objects by passing the values to a constructor with a parameter of the same name (much like JSON.net can). This would allow truly readonly properties, and keep the class simpler.

### Steps to reproduce
It would be preferable to support a class structured like the following:

```c#
public class SampleClass
{
    public string Prop1 { get; }
    public int Prop2 { get; }

    public SampleClass(string prop1, int prop2)
    {
        Prop1 = prop1;
        Prop2 = prop2;
    }
}
```

### Further technical details
EF Core version: 2.0.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer

**Comment by ajcvickers** [comment]
@codepb This is already implemented in the 2.1 code. You might consider giving it a try using our nightly builds, which are listed on the homepage: https://github.com/aspnet/EntityFrameworkCore

**Comment by codepb** [comment]
Excellent news, thanks! I didn't see an issue for this already, so apologies if I missed that.

**Comment by Tarig0** [comment]
Can we have Ef replace immutables, as described here?

Think currently with the use case you would have to manually detach the old entity and reattach the new entity.

**Comment by ajcvickers** [comment]
Note for triage: work still to be done here:
* Add fluent API for 
  * Choosing the constructor to use/ignore
  * Configuring a factory method instead
  * Binding to a service

**Comment by ajcvickers** [comment]
Closing this as fixed, with additional work tracked in #10789
